src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentPullDiagnosticHandler.cs

```diff
@@ -76,14 +77,16 @@ protected override ImmutableArray<Document> GetOrderedDocuments(RequestContext c
             return ImmutableArray.Create(context.Document);
         }
 
-        protected override Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(
+        protected override async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(
             RequestContext context, Document document, Option2<DiagnosticMode> diagnosticMode, CancellationToken cancellationToken)
         {
             // For open documents, directly use the IDiagnosticAnalyzerService.  This will use the actual snapshots
             // we're passing in.  If information is already cached for that snapshot, it will be returned.  Otherwise,
             // it will be computed on demand.  Because it is always accurate as per this snapshot, all spans are correct
             // and do not need to be adjusted.
-            return _analyzerService.GetDiagnosticsAsync(document.Project.Solution, documentId: document.Id, cancellationToken: cancellationToken);
```

Comment:
interesting - so GetDiagnosticsAsync will compute all project diagnostics when we pass a document?  I could see that causing slow diagnostics!

It might be useful to update the documentation on `GetDiagnosticsAsync` because it isn't exactly clear what scope this request applies to - right now it just says it gets diagnostics for a solution.

imho it would be nice if there was some way to clearly indicate the analysis scope directly from the API input, e.g. you could pass in a 'span scope' or a 'document scope' or a 'project scope' or a 'solution scope' and the diagnostics would be limited that way, but that may not be feasible.

Reply:
> interesting - so GetDiagnosticsAsync will compute all project diagnostics when we pass a document? I could see that causing slow diagnostics!

Yes.  That was surprising to me as well *especially* since the caller passes in a DocumentId to get results for.  However, the way it works *currently* is that is computes for the entire project then filters down to the doc.  I originally started by making it so it would only compute for the doc.  However i got nervous that through these very complex codepaths it might cache something making it say "i've computed for the entire project" and then a request for another document might return nothing.

I need @mavasani  to weigh in here on what teh right approach is and what we need to look out for.  

> imho it would be nice if there was some way to clearly indicate the analysis scope directly from the API input, e.g. you could pass in a 'span scope' or a 'document scope' or a 'project scope' or a 'solution scope' and the diagnostics would be limited that way, but that may not be feasible.

So the original API *does* take optional ProjectId/DocId, which (in a sense) does exactly that.  Except it works by doign the full computation and hten filtering down the results it computes to the Ids requested.  I'ts unclear to me if we can *easily* change the impl to use those IDs to limit what is computed due to how complex this code is.